book/html.md

```diff
@@ -8,205 +8,378 @@ next: layout
+very first thing in an HTML document. But it's not really an element
+at all, nor is it supposed to have a close tag. Our toy browser won't
+be using the doctype for anything, so it's best to throw it
+away:[^quirks-mode]
+
+[html5-doctype]: https://html.spec.whatwg.org/multipage/syntax.html#the-doctype
+
+[^quirks-mode]: Real browsers use doctypes to switch between
+    standards-compliant and legacy parsing and layout modes.
+
+``` {.python indent=4}
+def add_tag(self, tag):
+    if tag.startswith("!"): return
     # ...
-    def __repr__(self):
-        return "<" + self.tag + ">"
 ```
 
-This produces a more reasonable result:
+This ignores all tags that start with an exclamation mark, which not
+only throws out doctype declarations but also most comments, which in
+HTML are written `<!-- comment text -->`.
 
-``` 
-[<!DOCTYPE html>,
- <html lang="en-US" xml:lang="en-US">,
- <head>,
- <meta charset="utf-8" />,
- <meta name="generator" content="pandoc" />,
- <meta name="viewport" content=... />,
- <link rel="prev" href="text" />,
- <link rel="next" href="layout" />,
- <link rel="stylesheet" href="../book.css" />]
-```
+Just throwing out doctypes isn't quite enough though---if you run your
+parser now, it will crash. That's because after the doctype comes a
+newline, which our parser treats as text and tries to insert into the
+tree. Except there isn't a tree, since the parser hasn't seen any open
+tags. For simplicity, let's just have our browser skip whitespace-only
+text nodes to side-step the problem:[^ignore-them]
 
-Why aren't these open elements closed? Well, most of them (like
-`<meta>` and `<link>`) are what are called self-closing: you don't
-ever write `</meta>` or `</link>`. These tags don't need a close tag
-because they never surround content. Let's add that to our parser:
+[^ignore-them]: Real browsers retain whitespace to correctly render
+    `make<span></span>up` as one word and `make<span> </span>up` as
+    two. Our browser won't. Plus, ignoring whitespace simplifies
+    [later chapters](layout.md) by avoiding a special-case for
+    whitespace-only text tags.
 
+``` {.python indent=4}
+def add_text(self, text):
+    if text.isspace(): return
+    # ...
+```
 
-``` {.python indent=8 replace=parent)/parent%2c%20tok.attributes)}
-# ...
-elif tok.tag in SELF_CLOSING_TAGS:
-    node = ElementNode(tok.tag, parent)
-    parent.children.append(node)
+The parsed HTML tree now looks like this:
+
+``` {.example}
+<HTML LANG="EN-US" XML:LANG="EN-US">
+   <HEAD>
+     <META CHARSET="UTF-8" />
+       <META NAME="GENERATOR" CONTENT="PANDOC" />
+         <META NAME="VIEWPORT" CONTENT="WIDTH=DEVICE-WIDTH,INITIAL-SCALE=1.0,USER-SCALABLE=YES" />
+           <META NAME="AUTHOR" CONTENT="PAVEL PANCHEKHA &AMP; CHRIS HARRELSON" />
+             <LINK REL="STYLESHEET" HREF="BOOK.CSS" />
+               <LINK REL="STYLESHEET" HREF="HTTPS://FONTS.GOOGLEAPIS.COM/CSS?FAMILY=VOLLKORN%7CLORA&DISPLAY=SWAP" />
```

Comment:
URLs are case-sensitive. I think you will need to remove the upppercase code in the parser.